datalad/metadata/metadata.py

```diff
@@ -330,22 +366,73 @@ def __call__(
             # ignore submodule entries
             content = [ap for ap in content_by_ds[ds_path]
                        if ap.get('type', None) != 'dataset' or ap['path'] == ds_path]
-            if not content:
+            if not content and not show_keys:
                 # nothing other than subdatasets were given or discovered in
                 # this dataset, ignore
                 continue
             ds = Dataset(ds_path)
+            db_path = opj(ds.path, '.datalad', 'metadata', 'dataset.json')
+            db = {}
+            if exists(db_path):
+                db_fp = open(db_path)
+                # need to read manually, load() would puke on an empty file
+                db_content = db_fp.read()
+                # minimize time for collision
+                db_fp.close()
+                if db_content:
+                    db = json.loads(db_content)
+            #
+            # dedicated key handling
+            #
+            defs = db.get('definition', {})
+            if show_keys:
+                for k in sorted(defs):
+                    ui.message('{}: {} ({}: {})'.format(
+                        ac.color_word(k, ac.BOLD),
+                        defs[k],
+                        ac.color_word('dataset', ac.MAGENTA),
+                        ds.path))
+                # no other processing
+                continue
+            added_def = False
+            if define_key:
+                for k, v in define_key.items():
+                    if k in defs:
+                        if not defs[k] == v:
+                            yield get_status_dict(
+                                status='error',
+                                ds=ds,
+                                message=(
+                                    "conflicting definition for key '%s': '%s' != '%s'",
+                                    k, v, defs[k]),
+                                **res_kwargs)
+                            continue
+                    else:
+                        defs[k] = v
+                        added_def = True
+                db['definition'] = defs
+            #
+            # validate keys (only possible when dataset-defined keys are known
+            #
+            known_keys = set(common_key_defs.keys()).union(set(defs.keys()))
+            key_error = False
+            for cat in (init, add, reset) if not permit_undefined_keys else []:
+                for k in cat if cat else []:
+                    if k not in known_keys:
+                        yield get_status_dict(
+                            status='error',
+                            ds=ds,
+                            message=(
+                                "undefined key '%s', check spelling or use --define-key",
```

Comment:
Should we point to our bug tracker to suggest new ones?

Reply:
Fixed.